File: rgctl-tests/ecommerce-python/app/services/review.py
import uuid

from sqlalchemy.orm import Session

from app.exceptions import ConflictError, NotFoundError
from app.models.review import Review
from app.repositories.product import ProductRepository
from app.repositories.review import ReviewRepository
from app.schemas.review import CreateReviewRequest, ReviewResponse


class ReviewService:
    def __init__(self, db: Session) -> None:
        self.reviews = ReviewRepository(db)
        self.products = ProductRepository(db)

    def list_for_product(self, product_id: str) -> list[ReviewResponse]:
        if not self.products.get(product_id):
            raise NotFoundError("Product not found")
        reviews = self.reviews.list_for_product(product_id)
        return [ReviewResponse.model_validate(r) for r in reviews]

    def create(self, user_id: str, product_id: str, req: CreateReviewRequest) -> ReviewResponse:
        if not self.products.get(product_id):
            raise NotFoundError("Product not found")
        if self.reviews.find_by_user_and_product(user_id, product_id):
            raise ConflictError("You already reviewed this product")
        review = Review(
            id=str(uuid.uuid4()),
            product_id=product_id,
            user_id=user_id,
            rating=req.rating,
            comment=req.comment,
        )
        self.reviews.add(review)
        return ReviewResponse.model_validate(review)

File: rgctl-tests/ecommerce-go/internal/service/review.go
package service

import (
	"github.com/google/uuid"
	"github.com/sshaaf/ecommerce-go/internal/models"
	"github.com/sshaaf/ecommerce-go/internal/pkg/timeutil"
	"github.com/sshaaf/ecommerce-go/internal/repository"
)

type ReviewService struct {
	reviews  *repository.ReviewRepository
	products *repository.ProductRepository
}

func NewReviewService(reviews *repository.ReviewRepository, products *repository.ProductRepository) *ReviewService {
	return &ReviewService{reviews: reviews, products: products}
}

type CreateReviewRequest struct {
	Rating  int64  `json:"rating" binding:"required,min=1,max=5"`
	Comment string `json:"comment"`
}

type ReviewResponse struct {
	ID        string `json:"id"`
	ProductID string `json:"product_id"`
	UserID    string `json:"user_id"`
	Rating    int64  `json:"rating"`
	Comment   string `json:"comment"`
}

func (s *ReviewService) List(productID string) ([]ReviewResponse, error) {
	if _, err := s.products.FindByID(productID); err != nil {
		return nil, MapRepoError(err)
	}
	reviews, err := s.reviews.ListForProduct(productID)
	if err != nil {
		return nil, err
	}
	out := make([]ReviewResponse, len(reviews))
	for i, r := range reviews {
		out[i] = toReviewResponse(r)
	}
	return out, nil
}

func (s *ReviewService) Create(userID, productID string, req CreateReviewRequest) (*ReviewResponse, error) {
	if _, err := s.products.FindByID(productID); err != nil {
		return nil, MapRepoError(err)
	}
	review := &models.Review{
		ID:        uuid.NewString(),
		ProductID: productID,
		UserID:    userID,
		Rating:    req.Rating,
		Comment:   req.Comment,
		CreatedAt: timeutil.NowISO(),
	}
	if err := s.reviews.Create(review); err != nil {
		return nil, err
	}
	resp := toReviewResponse(*review)
	return &resp, nil
}

func toReviewResponse(r models.Review) ReviewResponse {
	return ReviewResponse{
		ID:        r.ID,
		ProductID: r.ProductID,
		UserID:    r.UserID,
		Rating:    r.Rating,
		Comment:   r.Comment,
	}
}
